File: training_data/reflection-tuning/day16_part2_2021.rb
class Packet
  attr_reader :version, :type_id, :value, :sub_packets

  def initialize(binary)
    @version = binary[0, 3].to_i(2)
    @type_id = binary[3, 3].to_i(2)
    @sub_packets = []
    @remaining = parse(binary[6..-1])
  end

  def parse(binary)
    if @type_id == 4
      parse_literal(binary)
    else
      parse_operator(binary)
    end
  end

  def parse_literal(binary)
    value_bits = ""
    i = 0
    loop do
      group = binary[i, 5]
      value_bits += group[1, 4]
      i += 5
      break if group[0] == '0'
    end
    @value = value_bits.to_i(2)
    binary[i..-1]
  end

  def parse_operator(binary)
    length_type_id = binary[0]
    if length_type_id == '0'
      length = binary[1, 15].to_i(2)
      sub_packets_binary = binary[16, length]
      parse_sub_packets(sub_packets_binary)
      binary[(16 + length)..-1]
    else
      num_sub_packets = binary[1, 11].to_i(2)
      remaining = binary[12..-1]
      num_sub_packets.times do
        packet = Packet.new(remaining)
        @sub_packets << packet
        remaining = packet.remaining
      end
      remaining
    end
  end

  def parse_sub_packets(binary)
    while binary.length > 0
      packet = Packet.new(binary)
      @sub_packets << packet
      binary = packet.remaining
    end
  end

  def version_sum
    @version + @sub_packets.sum(&:version_sum)
  end

  def evaluate
    case @type_id
    when 0 then @sub_packets.sum(&:evaluate)
    when 1 then @sub_packets.map(&:evaluate).reduce(:*)
    when 2 then @sub_packets.map(&:evaluate).min
    when 3 then @sub_packets.map(&:evaluate).max
    when 4 then @value
    when 5 then @sub_packets[0].evaluate > @sub_packets[1].evaluate ? 1 : 0
    when 6 then @sub_packets[0].evaluate < @sub_packets[1].evaluate ? 1 : 0
    when 7 then @sub_packets[0].evaluate == @sub_packets[1].evaluate ? 1 : 0
    end
  end

  def remaining
    @remaining
  end
end

hex = File.read('input.txt').strip
binary = hex.chars.map { |c| c.to_i(16).to_s(2).rjust(4, '0') }.join

packet = Packet.new(binary)
puts "Part 1: #{packet.version_sum}"
puts "Part 2: #{packet.evaluate}"

File: clojure/day16_part2_2021.clj

(ns packet-decoder
  (:require [clojure.string :as str]))

(def hex-to-bin
  {"0" "0000" "1" "0001" "2" "0010" "3" "0011"
   "4" "0100" "5" "0101" "6" "0110" "7" "0111"
   "8" "1000" "9" "1001" "A" "1010" "B" "1011"
   "C" "1100" "D" "1101" "E" "1110" "F" "1111"})

(defn hex-to-binary [hex-str]
  (->> hex-str
       (map str)
       (map hex-to-bin)
       (str/join "")))

(defn binary-to-long [binary-str]
  (Long/parseLong binary-str 2))

(defn parse-literal [binary-str]
  (loop [remaining binary-str
         value-str ""]
    (let [group (subs remaining 0 5)
          prefix (subs group 0 1)
          data (subs group 1 5)
          new-value-str (str value-str data)
          new-remaining (subs remaining 5)]
      (if (= prefix "0")
        {:value (binary-to-long new-value-str)
         :remaining new-remaining}
        (recur new-remaining new-value-str)))))

(defn parse-packet [binary-str]
  (let [version (binary-to-long (subs binary-str 0 3))
        type-id (binary-to-long (subs binary-str 3 6))]
    (if (= type-id 4)
      (let [{:keys [value remaining]} (parse-literal (subs binary-str 6))]
        {:version version
         :type-id type-id
         :value value
         :remaining remaining})
      (let [length-type-id (subs binary-str 6 7)]
        (if (= length-type-id "0")
          (let [total-length (binary-to-long (subs binary-str 7 22))
                sub-packets-str (subs binary-str 22 (+ 22 total-length))
                remaining (subs binary-str (+ 22 total-length))
                sub-packets (loop [remaining-sub sub-packets-str
                                   packets []]
                              (if (empty? remaining-sub)
                                packets
                                (let [parsed-packet (parse-packet remaining-sub)]
                                  (recur (:remaining parsed-packet)
                                         (conj packets parsed-packet)))))
                ]
            {:version version
             :type-id type-id
             :sub-packets sub-packets
             :remaining remaining})
          (let [num-sub-packets (binary-to-long (subs binary-str 7 18))
                remaining (subs binary-str 18)
                sub-packets (loop [remaining-sub remaining
                                   packets []
                                   count 0]
                              (if (or (empty? remaining-sub) (= count num-sub-packets))
                                packets
                                (let [parsed-packet (parse-packet remaining-sub)]
                                  (recur (:remaining parsed-packet)
                                         (conj packets parsed-packet)
                                         (inc count)))))
                ]
            {:version version
             :type-id type-id
             :sub-packets sub-packets
             :remaining (reduce (fn [acc p] (:remaining p)) remaining sub-packets)}))))))

(defn sum-versions [packet]
  (let [{:keys [version sub-packets]} packet]
    (if (empty? sub-packets)
      version
      (+ version (reduce + (map sum-versions sub-packets))))))

(defn calculate-value [packet]
  (let [{:keys [type-id value sub-packets]} packet]
    (case type-id
      0 (reduce + (map calculate-value sub-packets))
      1 (reduce * (map calculate-value sub-packets))
      2 (apply min (map calculate-value sub-packets))
      3 (apply max (map calculate-value sub-packets))
      4 value
      5 (if (> (calculate-value (first sub-packets)) (calculate-value (second sub-packets))) 1 0)
      6 (if (< (calculate-value (first sub-packets)) (calculate-value (second sub-packets))) 1 0)
      7 (if (= (calculate-value (first sub-packets)) (calculate-value (second sub-packets))) 1 0))))

(defn solve [input-str]
  (let [binary-str (hex-to-binary input-str)
        parsed-packet (parse-packet binary-str)]
    {:part1 (sum-versions parsed-packet)
     :part2 (calculate-value parsed-packet)}))

(defn -main []
  (let [input-str (slurp "input.txt")
        results (solve input-str)]
    (println "Part 1:" (:part1 results))
    (println "Part 2:" (:part2 results))))

(-main)
